### PR TITLE
Feature/implement statement value type

### PR DESCRIPTION
**Pull request by vinzlange**

### CODE REVIEW

File: src/attributes/AttributeValueTypes.ts
Comment (jkoenig134) [inline src/attributes/AttributeValueTypes.ts:1]: Statement missing in Classes and Type names

Comment (jkoenig134) [reply src/attributes/AttributeValueTypes.ts:1]: also xml has to be moved from there

File: src/attributes/types/statement/StatementIssuerConditions.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+export interface IStatementIssuerConditions extends IAbstractComplexValue {
+    validFrom: ICoreDate
+    validTo: ICoreDate
+    evidence: StatementEvidence
+    authorityType: StatementAuthorityType
+    relayedParty?: DigitalIdentityDescriptor
+}
+
+@type("statementIssuerConditions")
```
Comment (jkoenig134) [inline src/attributes/types/statement/StatementIssuerConditions.ts:25]:
```suggestion
@type("StatementIssuerConditions")
```

File: src/attributes/types/statement/AbstractStatement.ts
Comment (jkoenig134) [inline src/attributes/types/statement/AbstractStatement.ts:1]: missing valueHints and renderHints

File: src/attributes/types/statement/StatementIssuerConditions.ts
Comment (jkoenig134) [inline src/attributes/types/statement/StatementIssuerConditions.ts:1]: missing valueHints and renderHints

File: test/attributeValues/StatementValueTest.test.ts [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+import itParam from "mocha-param"
+import { AbstractTest } from "../AbstractTest"
+import { GenericValueTest } from "./GenericValueTest"
+
+export class StatementValueTest extends AbstractTest {
+    public run(): void {
+        this.runPerson()
+    }
+
+    public runPerson(): void {
```
Comment (jkoenig134) [inline test/attributeValues/StatementValueTest.test.ts:19]:
```suggestion
        this.runStatement()
    }

    public runStatement(): void {
```

**Review by jkoenig134 — APPROVED** [review]

**Review by jkoenig134 — CHANGES_REQUESTED** [review]
some small things

File: test/attributeValues/StatementValueTest.test.ts [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+                        value: "ownFact"
+                    },
+                    authorityType: {
+                        "@type": "StatementAuthorityType",
+                        value: "ownAuthority"
+                    }
+                })
+            }
+        })
+        describe("Statement negative Test", function () {
```
Comment (jkoenig134) [inline test/attributeValues/StatementValueTest.test.ts:238]:
```suggestion
        })

        describe("Statement negative Test", function () {
```

File: test/attributeValues/StatementValueTest.test.ts [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+            describe("Property: has Attribute without Attribute", function () {
```
Comment (jkoenig134) [inline test/attributeValues/StatementValueTest.test.ts:239]:
```suggestion
            describe("Statement set to 'hasPredicate'", function () {
```

File: test/attributeValues/StatementValueTest.test.ts [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+                            attributes: []
+                        },
+                        {
+                            address: "id1235"
+                        },
+                        {
+                            address: "id1235",
+                            attributes: ""
+                        },
+                        {}
```
Comment (jkoenig134) [inline test/attributeValues/StatementValueTest.test.ts:254]: this throws b/c of a different reason that you actually want to check. Please make sure that the correct error (`If the predicate of the ...`) is thrown.

File: src/attributes/types/statement/AbstractStatement.ts [hunk 6]
```diff
@@ -78,4 +79,15 @@ export abstract class AbstractStatement extends AbstractComplexValue implements
     ): AbstractStatementJSON {
         return super.toJSON(verbose, serializeAsString) as AbstractStatementJSON
     }
+
+    public static get valueHints(): ValueHints {
+        return ValueHints.from({})
```
Comment (jkoenig134) [inline src/attributes/types/statement/AbstractStatement.ts:84]: missing propertyHints

File: src/attributes/types/statement/StatementIssuerConditions.ts [hunk 7]
```diff
@@ -56,4 +57,14 @@ export class StatementIssuerConditions extends AbstractComplexValue implements I
     ): StatementIssuerConditionsJSON {
         return super.toJSON(verbose, serializeAsString) as StatementIssuerConditionsJSON
     }
+    public static get valueHints(): ValueHints {
+        return ValueHints.from({})
```
Comment (jkoenig134) [inline src/attributes/types/statement/StatementIssuerConditions.ts:61]: missing propertyHints

File: test/attributeValues/StatementValueTest.test.ts [hunk 8]
```diff
@@ -275,7 +272,10 @@ export class StatementValueTest extends AbstractTest {
                                 }
                             }
                         }
-                        expect(() => Statement.fromAny(invalidStatement)).to.throw()
+                        expect(() => Statement.fromAny(invalidStatement)).to.throw(
+                            ValidationError,
+                            "If the predicate of the Function is 'hasAttribute' you have to define attributes in 'object.attributes'"
```
Comment (jkoenig134) [inline test/attributeValues/StatementValueTest.test.ts:277]: "If the predicate of the _**Function**_" ?!?!?! Seems like something isn't right here.

File: src/attributes/types/statement/AbstractIdentityDescriptor.ts [hunk 9]
```diff
@@ -14,4 +15,15 @@ export abstract class AbstractIdentityDescriptor extends Serializable implements
     @serialize()
     @validate({ nullable: true })
     public attributes?: IdentityAttribute[]
+
+    public static get valueHints(): ValueHints {
+        return ValueHints.from({})
```
Comment (jkoenig134) [inline src/attributes/types/statement/AbstractIdentityDescriptor.ts:20]: missing propertyHints

File: src/attributes/types/statement/AbstractStatement.ts [hunk 10]
```diff
@@ -81,13 +81,26 @@ export abstract class AbstractStatement extends AbstractComplexValue implements
     }
 
     public static get valueHints(): ValueHints {
-        return ValueHints.from({})
+        return ValueHints.from({
```
[the remark this answers is not in the document]
Comment (vinzlange) [reply src/attributes/types/statement/AbstractStatement.ts:84]: Property 'valueHints' does not exist on type 'typeof AbstractComplexValue'

File: src/attributes/types/statement/AbstractIdentityDescriptor.ts [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    public static readonly propertyNames: any = nameOf<AbstractIdentityDescriptor, never>()
+
+    @serialize()
+    @validate({ nullable: true })
+    public attributes?: IdentityAttribute[]
+
+    public static get valueHints(): ValueHints {
+        return ValueHints.from({
+            propertyHints: {
+                [this.propertyNames.subject.$path]: ValueHints.from({})
```
Comment (jkoenig134) [inline src/attributes/types/statement/AbstractIdentityDescriptor.ts:29]: subject?!

File: src/attributes/types/statement/AbstractIdentityDescriptor.ts [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+            }
+        })
+    }
+
+    public static override get renderHints(): RenderHints {
+        return super.renderHints.copyWith({
+            propertyHints: {
+                [this.propertyNames.recipient.$path]: RenderHints.from({
```
Comment (jkoenig134) [inline src/attributes/types/statement/AbstractIdentityDescriptor.ts:37]: recipient?!

File: src/attributes/types/statement/AbstractStatement.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (jkoenig134) [reply src/attributes/types/statement/AbstractStatement.ts:84]: I see, only renderHints are supported.

File: src/attributes/types/statement/StatementIssuerConditions.ts [hunk 13]
```diff
@@ -58,13 +61,32 @@ export class StatementIssuerConditions extends AbstractComplexValue implements I
         return super.toJSON(verbose, serializeAsString) as StatementIssuerConditionsJSON
     }
     public static get valueHints(): ValueHints {
```
Comment (jkoenig134) [inline src/attributes/types/statement/StatementIssuerConditions.ts:63]:
```suggestion
    }

    public static get valueHints(): ValueHints {
```

**Review by jkoenig134 — APPROVED** [review]

**Pull request merged**